test/jdk/java/awt/regtesthelpers/PassFailJFrame.java

```diff
@@ -314,6 +303,54 @@ public interface WindowCreator {
     }
 
 
+    private static final class TimeoutHandler implements ActionListener {
+        private final long endTime;
+
+        private final Timer timer;
+
+        private final JLabel label;
+
+        public TimeoutHandler(final JLabel label, final long testTimeOut) {
+            endTime = System.currentTimeMillis() + TimeUnit.MINUTES.toMillis(testTimeOut);
+
+            this.label = label;
+
+            timer = new Timer(1000, this);
+            timer.start();
+            updateTime(testTimeOut);
+        }
+
+        @Override
+        public void actionPerformed(ActionEvent e) {
+            long leftTime = endTime - System.currentTimeMillis();
+            if ((leftTime < 0) || failed) {
+                timer.stop();
+                testFailedReason = FAILURE_REASON
+                                   + "Timeout User did not perform testing.";
```

Comment:
If the test was explicitly failed by the user (ie failed==true), why does the reason say that there was a time out ?

Reply:
Huh, it's a good question. I didn't notice it. It's the original code that I moved into a dedicated class.

It shouldn't change the failure reason if it's already set.

The timer isn't stopped in handlers for the failure, so it's possible that timer changes it. We should also stop the timer explicitly when failure or success is handled.

Reply:
@prrace Eventually, I refactored handling of `failureReason`.  I removed `failed` and `timeout` fields; the `failureReason` is set via a synchronised method to ensure there cannot be data race. If `failureReason` is not `null`, the test fails. After `failureReason` is set to a non-`null` value, it cannot be changed.

The timeout handler relies on this behaviour: it calls `setFailureReason` unconditionally; if `failureReason` is already set, the timeout reason is ignored.